### PR TITLE
Add scenarios for route handler filters

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/dotnet/aspnetcore/issues/40655

### CODE REVIEW

File: build/trend-scenarios.yml [hunk 1]
```diff
@@ -38,11 +38,16 @@ parameters:
     arguments: --scenario mapaction $(plaintextJobs) --property scenario=PlaintextMapAction --property protocol=http
   - displayName: Plaintext Endpoint
     arguments: --scenario endpoint $(plaintextJobs) --property scenario=PlaintextEndpoint --property protocol=http
+  - displayName: Plaintext (No Filters)
+    arguments: --scenario plainTextSansFilter $(plaintextJobs) --property scenario=PlaintextWithParameterNoFilter --property protocol=http
+  - displayName: Plaintext (Empty Filters)
+    arguments: --scenario plainTextEmptyFilter $(plaintextJobs) --property scenario=PlaintextWithParameterEmptyFilter --property protocol=http
+  - displayName: Plaintext (With Filters)
+    arguments: --scenario plainTextWithFilter $(plaintextJobs) --property scenario=PlaintexWithParametertWithFilter --property protocol=http
```
Comment (BrennanConroy) [inline build/trend-scenarios.yml:46]:
```suggestion
    arguments: --scenario plainTextWithFilter $(plaintextJobs) --property scenario=PlaintexWithParameterWithFilter --property protocol=http
```

File: src/BenchmarksApps/MapAction/Program.cs [hunk 2]
```diff
@@ -23,6 +25,29 @@
...
                 object Json() => new { message = "Hello, World!" };
                 endpoints.MapGet("/json", (Func<object>)Json);
+
+                // Parameterized plain-text endpoint
+                string SayHello(string name) => $"Hello, {name}!";
+                // With no filters
+                endpoints.MapGet("/hello/{name}", SayHello);
+                // With a filter that no-ops
+                endpoints.MapGet("/helloEmptyFilter/{name}", SayHello)
+                    .AddFilter(async (context, next) => await next(context));
```
Comment (BrennanConroy) [inline src/BenchmarksApps/MapAction/Program.cs:35]:
```suggestion
                    .AddFilter((context, next) => next(context));
```

File: src/BenchmarksApps/MapAction/Program.cs [hunk 3]
```diff
@@ -23,6 +25,29 @@
...
+                        var hasSingleParam = routeHandlerContext.MethodInfo.GetParameters().Length == 1;
+                        var isStringParam = routeHandlerContext.MethodInfo.GetParameters()[0].ParameterType == typeof(string);
+
+                        return async (context) =>
+                        {
+                            if (hasSingleParam && isStringParam)
+                            {
+                                context.Parameters[0] = context.Parameters[0].ToUpperInvariant();
+                            }
+                            return await next(context);
```
Comment (BrennanConroy) [inline src/BenchmarksApps/MapAction/Program.cs:48]:
```suggestion
                        return (context) =>
                        {
                            if (hasSingleParam && isStringParam)
                            {
                                context.Parameters[0] = context.Parameters[0].ToUpperInvariant();
                            }
                            returnnext(context);
```

File: src/BenchmarksApps/MapAction/Program.cs [hunk 4]
```diff
@@ -23,6 +25,29 @@
...
+
+                        return (context) =>
+                        {
+                            if (hasSingleParam && isStringParam)
+                            {
+                                context.Parameters[0] = context.Parameters[0].ToUpperInvariant();
+                            }
+                            return next(context);
+                        };
+                    });
```
Comment (davidfowl) [inline src/BenchmarksApps/MapAction/Program.cs:50]: What is this one for?

Comment (davidfowl) [reply src/BenchmarksApps/MapAction/Program.cs:50]: I see parameter mutation?

Comment (captainsafia) [reply src/BenchmarksApps/MapAction/Program.cs:50]: Yeah, we can. I was trying to do the base case (no filters), pass through case (filter that does nothing but invoke next), and a case with some logic. I guess what logic we want to test will depend on what we care about.

File: src/BenchmarksApps/MapAction/Program.cs [hunk 5]
```diff
@@ -27,21 +27,24 @@
...
 
                 // Parameterized plain-text endpoint
-                string SayHello(string name) => $"Hello, {name}!";
+                string SayHello(string name, int age, string location) => $"Hello, {name}! You're {age} years old and based in {location}.";
                 // With no filters
-                endpoints.MapGet("/hello/{name}", SayHello);
+                endpoints.MapGet("/hello/{name}/{age}/{location}", SayHello);
                 // With a filter that no-ops
-                endpoints.MapGet("/helloEmptyFilter/{name}", SayHello)
+                endpoints.MapGet("/helloEmptyFilter/{name}/{age}/{location}", SayHello)
```
Comment (davidfowl) [inline src/BenchmarksApps/MapAction/Program.cs:34]: Does this work? We need  anew handler no?

Comment (captainsafia) [reply src/BenchmarksApps/MapAction/Program.cs:34]: The `plaintextNoParamsWithFilter` validates a pass through filter without parameters. This one validates a pass through filter with parameters.

Comment (davidfowl) [reply src/BenchmarksApps/MapAction/Program.cs:34]: But the method doesn't have the arguments. Can we add another handler?

Comment (halter73) [reply src/BenchmarksApps/MapAction/Program.cs:34]: What do you mean the method doesn't have arguments? Are you referring to `SayHello`?

File: src/BenchmarksApps/MapAction/Program.cs [hunk 6]
```diff
@@ -23,6 +25,32 @@
...
+                    .AddFilter((context, next) => next(context));
+                endpoints
+                    .MapGet("/helloFiltered/{name}/{age}/{location}", SayHello)
+                    .AddFilter((routeHandlerContext, next) => {
+                        var hasParams = routeHandlerContext.MethodInfo.GetParameters().Length >= 1;
+                        var isStringParam = routeHandlerContext.MethodInfo.GetParameters()[0].ParameterType == typeof(string);
+
+                        return (context) =>
+                        {
+                            if (hasParams && isStringParam)
```
Comment (halter73) [inline src/BenchmarksApps/MapAction/Program.cs:47]: Nit: This if could be put outside the `return (context) => ...`

File: scenarios/plaintext.benchmarks.yml [hunk 7]
```diff
@@ -92,6 +92,46 @@ scenarios:
...
+    application:
+      job: mapaction
+    load:
+      job: wrk
+      variables:
+        presetHeaders: plaintext
+        path: /helloEmptyFilter/TesterName/42/Internet
+        pipeline: 16
+  
+  plainTextNoParamsEmptyFilter:
```
Comment (halter73) [inline scenarios/plaintext.benchmarks.yml:116]: Is the non-filter variant of this is just the `mapaction` scenario?

**Review by halter73 — APPROVED** [review]

File: scenarios/plaintext.benchmarks.yml (hunk 7 above)
Comment (captainsafia) [reply scenarios/plaintext.benchmarks.yml:116]: Yep, it's just a call to the `Plaintext` endpoint that spits out 'Hello world'

File: src/BenchmarksApps/MapAction/Program.cs (hunk 5 above)
Comment (captainsafia) [reply src/BenchmarksApps/MapAction/Program.cs:34]: David and I synced about it only. He thought I was reusing the `SayHello` handler for the `plainTextNoParams` scenarios.

**Pull request merged**